torch/_inductor/codegen/cpp.py

```diff
@@ -2281,21 +2289,38 @@ def _get_store_line(
                 code.writeline(f"{value}.store({var_expr}, {self.tiling_factor});")
         else:
             self._load_or_store_non_contiguous(
-                var, index, dtype, buffer=code, store_value=value
+                var, index, dtype, buffer=code, store_value=value, accu=accu
             )
         return code
 
     def store(self, name, index, value, mode=None):
         assert "buf" in name
-        assert mode is None
         assert isinstance(value, CppCSEVariable), value
         if not value.is_vec:
             # this happens when we store a scalar into a vectorized buffer like "fill"
             value = self.broadcast(value)
         var = self.args.output(name)
         index = self.rename_indexing(index)
-        code = self._get_store_line(value, var, index, V.graph.get_dtype(name))
-        self.stores.splice(code.map(lambda x: DeferredLine(name, x)))
+        if mode is None:
+            code = self._get_store_line(value, var, index, V.graph.get_dtype(name))
+            self.stores.splice(code.map(lambda x: DeferredLine(name, x)))
+        elif mode == "atomic_add":
+            if not config.cpp.dynamic_threads and self.num_threads == 1:
+                load_var = self.load_to_buffer(name, index, self.stores)
+                code = self._get_store_line(
+                    f"({load_var} + {value})",
+                    var,
+                    index,
+                    V.graph.get_dtype(name),
+                    accu=True,
+                )
+                self.stores.splice(code.map(lambda x: DeferredLine(name, x)))
+            else:
+                index = ops.index_expr(index, torch.int64)
```

Comment:
I guess the intention is to vectorize the index? But it is still possible that the index is a scalar in which case we should have a specialized version of `atomic_add_vec` for it?

Reply:
The intension here is to reuse `index_expr` to load contiguous or not contiguous index.
Although `index_expr` support to handle scalar index,  I don't think the index here should not be a scalar 
The original code use `_get_store_line` here and looks like `_get_store_line` assume index will not be a scalar:
```
    def _get_store_line(
        self,
        value: Union[str, CppCSEVariable],
        var: str,
        index: sympy.Expr,
        dtype: torch.dtype,
        accu_store: bool = False,
    ):
        """
        Get a store line buffer that stores `value` into `var` at `index` of `dtype`. It handles
        both contiguous and non-contiguous store cases.
        :param value: Vectorized type templaterized on `dtype`.
        :param var: buffer to store into.
        :index: index into the `var`.
        """
        # when value's type is str (e.g., welford reduction), caller should make sure
        # it is a vector
        assert isinstance(value, str) or (
            isinstance(value, CppCSEVariable) and value.is_vec
        ), value
        tiling_var = self.itervars[self.tiling_idx]
        var_expr = f"{var} + {cexpr_index(index)}"
        stride = self._try_get_const_stride(index, tiling_var)
        code = IndentedBuffer()
        if stride == 1:
            if dtype == torch.float and self.tail_size is None:
                code.writeline(f"{value}.store({var_expr});")
            else:
                code.writeline(f"{value}.store({var_expr}, {self.num_elems});")
        else:
            self._load_or_store_non_contiguous(
                var, index, dtype, buffer=code, store_value=value, accu_store=accu_store
            )
        return code
```

Reply:
If so, please add an assertion here to make sure the returned index is a vector: `assert index.is_vec`